File: sdk/python/examples/63_deploy.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Deploy — register agents on the server (CI/CD step).

Demonstrates:
    - runtime.deploy() to compile and register multiple agents
    - DeploymentInfo result with workflow name and agent name
    - CI/CD use case: push agent definitions without executing them

deploy() sends agent configs to the server, which compiles them into
Conductor workflow definitions and registers the corresponding task
definitions. No local workers are started, no execution happens.

Run this once during deployment. Use serve() separately (63b) to keep
workers alive, and run-by-name (63c) to trigger execution.

Requirements:
    - Conductor server running
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api in .env or environment
    - AGENT_LLM_MODEL=openai/gpt-4o-mini in .env or environment
"""

from agentspan.agents import Agent, AgentRuntime, tool
from settings import settings


@tool
def search_docs(query: str) -> str:
    """Search internal documentation.

    Args:
        query: Search query string.

    Returns:
        Matching documentation excerpts.
    """
    return f"Found 3 results for: {query}"


@tool
def check_status(service: str) -> str:
    """Check service health status.

    Args:
        service: Name of the service to check.

    Returns:
        Health status string.
    """
    return f"{service}: healthy"


# ── Define agents ────────────────────────────────────────────────────

doc_assistant = Agent(
    name="doc_assistant",
    model=settings.llm_model,
    tools=[search_docs],
    instructions="Help users find documentation. Use search_docs to look up answers.",
)

ops_bot = Agent(
    name="ops_bot",
    model=settings.llm_model,
    tools=[check_status],
    instructions="Monitor service health. Use check_status to inspect services.",
)

# ── Deploy: compile + register on server ─────────────────────────────

with AgentRuntime() as runtime:
    results = runtime.deploy(doc_assistant, ops_bot)

    for info in results:
        print(f"Deployed: {info.agent_name} -> {info.workflow_name}")

    print(f"\n{len(results)} agent(s) registered on server.")
    print("Now run 63b_serve.py to start workers, then 63c_run_by_name.py to execute.")

File: sdk/typescript/examples/63-deploy.ts
/**
 * 63 - Deploy — register agents on the server (CI/CD step).
 *
 * deploy() sends agent configs to the server, which compiles them into
 * Conductor workflow definitions. No local workers are started.
 *
 * Requirements:
 *   - Conductor server running
 *   - AGENTSPAN_SERVER_URL=http://localhost:6767/api as environment variable
 *   - AGENTSPAN_LLM_MODEL=openai/gpt-4o-mini as environment variable
 */

import { z } from 'zod';
import { Agent, AgentRuntime, tool } from '../src/index.js';
import { llmModel } from './settings.js';

// -- Tools -------------------------------------------------------------------

const searchDocs = tool(
  async (args: { query: string }) => {
    return `Found 3 results for: ${args.query}`;
  },
  {
    name: 'search_docs',
    description: 'Search internal documentation.',
    inputSchema: z.object({
      query: z.string().describe('Search query string'),
    }),
  },
);

const checkStatus = tool(
  async (args: { service: string }) => {
    return `${args.service}: healthy`;
  },
  {
    name: 'check_status',
    description: 'Check service health status.',
    inputSchema: z.object({
      service: z.string().describe('Name of the service to check'),
    }),
  },
);

// -- Define agents -----------------------------------------------------------

export const docAssistant = new Agent({
  name: 'doc_assistant',
  model: llmModel,
  tools: [searchDocs],
  instructions: 'Help users find documentation. Use search_docs to look up answers.',
});

export const opsBot = new Agent({
  name: 'ops_bot',
  model: llmModel,
  tools: [checkStatus],
  instructions: 'Monitor service health. Use check_status to inspect services.',
});

// -- Deploy: compile + register on server ------------------------------------

// Only run when executed directly (not when imported for discovery)
if (process.argv[1]?.endsWith('63-deploy.ts') || process.argv[1]?.endsWith('63-deploy.js')) {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(docAssistant, 'How do I reset my password?');
    result.printResult();

    // Production pattern:
    // 1. Deploy once during CI/CD:
    // await runtime.deploy(docAssistant);
    // await runtime.deploy(opsBot);
    // CLI alternative:
    // agentspan deploy --package sdk/typescript/examples
    //
    // 2. In a separate long-lived worker process:
    // await runtime.serve(docAssistant, opsBot);
  } finally {
    await runtime.shutdown();
  }
}
